File: android/digia-engage/src/main/java/com/digia/engage/internal/VariableContext.kt
package com.digia.engage.internal

internal data class VariableContext(
    val values: Map<String, String>,  // pre-resolved: CEP → fallback → ""
    val types: Map<String, String>,   // name → "string"|"number"
) {
    companion object {
        val empty = VariableContext(emptyMap(), emptyMap())
    }
}

/** Builds a [VariableContext] from schemas, letting non-empty CEP values win. */
internal fun buildVariableContext(
    schemas: List<VariableSchema>,
    cepVars: Map<String, String>?,
): VariableContext {
    val values = mutableMapOf<String, String>()
    val types  = mutableMapOf<String, String>()
    for (schema in schemas) {
        val cepVal = cepVars?.get(schema.name)
        values[schema.name] = if (!cepVal.isNullOrEmpty()) cepVal else schema.fallbackValue
        types[schema.name]  = schema.type
    }
    return VariableContext(values = values, types = types)
}

File: react-native/src/variable-context.ts
import type { VariableSchema } from './variable-schema';

export type VariableContext = {
    values: Record<string, string>;
    types: Record<string, 'string' | 'number'>;
};

// CEP → fallbackValue → "". A non-empty CEP value wins; empty falls through.
export const buildVariableContext = (
    schemas: VariableSchema[],
    cepVars: Record<string, string> | undefined,
): VariableContext => {
    const values: Record<string, string> = {};
    const types: Record<string, 'string' | 'number'> = {};

    for (const schema of schemas) {
        const cep = cepVars?.[schema.name];
        values[schema.name] = (cep !== undefined && cep !== '') ? cep : schema.fallbackValue;
        types[schema.name] = schema.type;
    }

    return { values, types };
};
